### ISSUE TITLE
ErrorCode for Checksum integrity

### ISSUE THREAD

**Issue by tommyziegler** [tatsuhiro-t/aria2#379]
When I pass the checksum via the RPC Interface I receive as an Error-Code value 1 back. (In case the Check fails)

Value 1 is like described in the documentation an unknown error. Is there an other way to find out if the checksum isn't correct?
How easy it would be to integrate that?

Greetings

Tommy Ziegler

**Comment by tatsuhiro-t** [comment]
Try with aria2c in command-line, and see what error it says.
Probably, the format is not right, or given message digest type is not supported.
Could you tell us how you set that option in RPC method?

**Comment by tommyziegler** [comment]
The Aria Command line says also Error-Code value 1. (Unknown error in Documentation)

See Terminal:
```
[12:45][TZiegler@tzieglers-mbp:~]$ aria2c --checksum=sha-256=CA11AEF69E65FBF666F6A5ABC73DCC9CCC28596642D52A2415965A319A190574 https://s3.eu-central-1.amazonaws.com/siu-large-download-test/jdk-8u45-windows-x64.exe

05/11 12:47:00 [NOTICE] Allocating disk space. Use --file-allocation=none to disable it. See --file-allocation option in man page for more details.
 *** Download Progress Summary as of Mon May 11 12:48:00 2015 ***                                                                                   
====================================================================================================================================================
[#bce7e8 130MiB/180MiB(72%) CN:1 DL:2.3MiB ETA:21s]
FILE: /Users/TZiegler/jdk-8u45-windows-x64.exe
----------------------------------------------------------------------------------------------------------------------------------------------------

[#bce7e8 180MiB/180MiB(100%) CN:0] [Checksum:#bce7e8 157MiB/180MiB(87%)]                                                                            
05/11 12:48:23 [NOTICE] Verification finished successfully. file=/Users/TZiegler/jdk-8u45-windows-x64.exe

05/11 12:48:23 [NOTICE] Download complete: /Users/TZiegler/jdk-8u45-windows-x64.exe

Download Results:
gid   |stat|avg speed  |path/URI
======+====+===========+=======================================================
bce7e8|OK  |   2.2MiB/s|/Users/TZiegler/jdk-8u45-windows-x64.exe

Status Legend:
(OK):download completed.
[12:48][TZiegler@tzieglers-mbp:~]$ echo $?
0
[12:48][TZiegler@tzieglers-mbp:~]$ aria2c --checksum=sha-256=CA11AEF69E65FBF666F6A5ABC73DCC9CCC28596642D52A2415965A319A190573 https://s3.eu-central-1.amazonaws.com/siu-large-download-test/jdk-8u45-windows-x64.exe

05/11 12:48:37 [NOTICE] File already exists. Renamed to /Users/TZiegler/jdk-8u45-windows-x64.exe.1.

05/11 12:48:37 [NOTICE] Allocating disk space. Use --file-allocation=none to disable it. See --file-allocation option in man page for more details.
 *** Download Progress Summary as of Mon May 11 12:49:37 2015 ***                                                                                   
====================================================================================================================================================
[#3cab33 139MiB/180MiB(77%) CN:1 DL:2.5MiB ETA:16s]
FILE: /Users/TZiegler/jdk-8u45-windows-x64.exe.1
----------------------------------------------------------------------------------------------------------------------------------------------------

[#3cab33 180MiB/180MiB(100%) CN:0] [Checksum:#3cab33 140MiB/180MiB(77%)]                                                                            
05/11 12:49:58 [ERROR] Checksum error detected. file=/Users/TZiegler/jdk-8u45-windows-x64.exe.1

05/11 12:49:58 [NOTICE] Download GID#3cab338af41ae41b not complete: /Users/TZiegler/jdk-8u45-windows-x64.exe.1

Download Results:
gid   |stat|avg speed  |path/URI
======+====+===========+=======================================================
3cab33|ERR |   2.3MiB/s|/Users/TZiegler/jdk-8u45-windows-x64.exe.1

Status Legend:
(ERR):error occurred.

aria2 will resume download if the transfer is restarted.
If there are any errors, then see the log file. See '-l' option in help/man page for details.
[12:49][TZiegler@tzieglers-mbp:~]$ echo $?
1
```

I use the option checksum with correct type when I add the URI to Aria. That works also if the checksum is correct. When the checksum is not correct Aria tells me over the info API that there was an error, and if I ask aria2 via the API for the error code I receive the same error like over the command line.

```checksum=sha-256=CA11AEF69E65FBF666F6A5ABC73DCC9CCC28596642D52A2415965A319A190573```


But I think when you have this checking possibility, for me as an user it's important to know that the error was the wrong checksum. Or what's your view to that?


Thanks and many greetings

Tommy

**Comment by tatsuhiro-t** [comment]
OK, so you are talking about exit status.  I misunderstood that RPC method returns error code when you specified checksum option.

I agree that specific error code for checksum validation failure is useful.
Currently, we don't have such category in error code, so we just return unknown error.

**Comment by tommyziegler** [comment]
Yes exactly. But I ask for the error code via the RPC API.

Do you have any idea how hard it would be to solve? Or where in the source this mapping is done that I could fork it or make an pull request?

Thx Tommy

**Comment by tatsuhiro-t** [comment]
Technically no so difficult.  We first add error code to error_code.h, then RequestGroup::setLastErrorCode(code) to set error code.  Calling place probably is in CheckIntegrityCommand.cc when checksum validation failed.
And write the document.

**Comment by tatsuhiro-t** [comment]
Fixed committed via 3855205.
Please test latest master.

**Comment by antbryan** [comment]
any feedback @tommyziegler ?
